packages/block-library/src/cover/index.php

```diff
@@ -60,10 +60,13 @@ class="wp-block-cover__image-background"
 			esc_attr( $object_position )
 		);
 
-		$content = str_replace(
-			'</span><div',
-			'</span>' . $image . '<div',
-			$content
+		/* Inserts the featured image between the (1st) cover 'background' `span` and 'inner_container' `div`,
+		   and removes eventual withespace characters between the two (typically introduced at template level) */
+		$content = preg_replace(
+		    '/(<span\b[^>]*wp-block-cover__background[\s|"][^>]*>.*<\/span>)\s*(<div\b[^>]*wp-block-cover__inner-container[\s|"][^>]*>)/U',
```

Comment:
"Processing" HTML with regex sucks no matter how you look at it. The more complex the regex, the more it sucks :) 

This looks very fragile. Is there any chance at all to do it in a more compatible and stable manner? One thing that comes to mind is to prepend or append the <img> tag, not attempt to insert it in the middle of some HTML. Trying to "splice" it as the current code does would most likely keep introducing edge cases, will likely break if the block is ever changed by a plugin, and will require ongoing maintenance.

If prepending/appending is impossible, seems there are other options:
1. Insert an HTML element as a placeholder in the editor. That way the specific location of the featured image can be determined much better, and the replacement code on the PHP side will be a bit more "bulletproof".
2. Reconstruct the block's HTML in PHP, i.e. make it really a dynamic block.

Reply:
One note on this regex: there's a major jump in required complexity between matching an HTML tag and parsing HTML. Parsing HTML tags is simpler and possible with traditional RegExp patterns and knowledge, but what we've done by attempting to capture the `<span>` and the `<div>` together is parsing full HTML. A trivial example input shows how [this can break down](https://regex101.com/r/VHvK8B/1):

```
<span class="wp-block-cover__background"><span class="wp-block-cover__background">Something else</span><div class="wp-block-cover__inner-container">Something</span><div class="wp-block-cover__inner-container">
```

The `.*` really is problematic, and you probably don't want to have to maintain an appropriate RegExp pattern to do this properly. There's no good way to match that specific `</span>` with a RegExp, which is why in my example I matched the `<div>` with its class name, because that's very doable. That is, if we are trying to identify a specific tag based on some attribute, we can only do that by matching the opening tag (or a void tag if there are no children). If we want to match the _end_ of a tag we only have a good option to match "any closing tag of this kind."

Also curious, why do we want to drop the match on the `class` attribute?

Reply:
@dmsnell 
1. Abut your curisity, take a look: https://github.com/WordPress/gutenberg/blob/51db4bf888e6b18cf9d18266108114d61070f3ad/lib/block-supports/layout.php#L220-L234
    
    As you can see, no `class=` to match the div with `class="wp-block-group"`, and the div with`class="wp-block-group__inner-container"` 
    
    And again: https://github.com/WordPress/gutenberg/blob/51db4bf888e6b18cf9d18266108114d61070f3ad/lib/block-supports/layout.php#L243-L246
    
    No `class=`. And the matching of a closing tag. And `.*` between the two.
    
    As in our case, they had to put something between the two groups.

2. With your **HTML malformed** every day usecase, putting a `div` inside a `span`:
  ```html
  <span><span>...</span><div>...</span><div>...
  ``` 
  my regex would insert the featured image before the last `div`. While the regex you suggested [before the first `div`](https://regex101.com/r/VHvK8B/1).
    So what?

It would seem you are maintaining at the same time:
- "It's a bad practice trying to cotrol what comes before the opening div. That's why I was suggesting no control at all"
- "I can't accept we are so permissive about what comes (within the span) before the div.We must absolutly control it"

Reply:
>0) One thing that comes to mind is to prepend or append the <img> tag, not attempt to insert it in the middle of some HTML.

I could be wrong, but I bet it could accomplished rewriting less than 10% of the entire Gutenberg!

> 1) Insert an HTML element as a placeholder in the editor.

Easy: rewite the Cover Block (and probably the Image Block)!

2) > Reconstruct the block's HTML in PHP

More or less this is what has been done by @draganescu. Who has added such a cool feature to WP. I was so impacient that installed Gutenberg for the first time to try it. And I had the misfortune to discover some kind of bug. It's not a real bug. And it wasn't me to discover the nature of this pseudo 'bug'. We have to remove some spaces and line breaks. That's it. If yu have some code, let's do it!

Reply:
@navigatrum first of all I want to say thank for working hard on this. please know that when I look at a proposed change my purpose and goal is to try and find the places where it could break down and so if I seem pessimistic it's because that's how I'm looking at the changes (and all changes for any PR).

Concerning the `class` question would it be right to understand that you are saying you removed it because there was other code in the project that also doesn't? If so, I wouldn't suggest we need to feel bound to match that. I'm not sure why it was left out there either, and it's not a critical thing, but it just seems surprising that we'd want to intentionally decouple the class name we're looking for from the `class` attribute we expect to find it in. 🤷‍♂️ 

> my regex would insert the featured image before the last div. While the regex you suggested [before the first div](https://regex101.com/r/VHvK8B/1).
So what?

I might not have presented a great example because it looks malformed. In [this example](https://regex101.com/r/LCAVkP/1) it should be more apparent that the `.*` will continue consuming input as long as it can while still matching the entire thing. Even a small change there could prevent a match from continuing past any unexpected tags (for example, if we use `[^<]*` instead of `.*` then we at least shouldn't break past unexpected tags that might sit between the `<span>` and the `<div>`).

We're probably somewhat shielded here because we're only being supplied the HTML from within the block itself (and also because we added the match on the class name), but if this were applied to the post (or if we were only looking at the tag names) we could easily get into trouble.

> It would seem you are maintaining at the same time:

I think there's a misunderstand here. What I'm saying is that we should be aware that whitespace isn't significant in HTML in the way we were treating it and we should look for an approach that doesn't depend on a specific whitespace formatting. We have very few guarantees about how it will reach our code.

When moving to the more robust RegExp there was simply another point I wanted to raise about the risk `.*` introduces and how it can cause our matcher to accidentally consume significantly more than we expect it to (and perform much slower while doing it!)

It's unfortunate we have more cases where we've already introduced these risks but I believe that's a good enough reason to be on alert to help spot these before they get merged. RegExp patterns can be hard enough, let alone with the complexity of trying to parse HTML…

---

In summary I think there's several ways to move this forward without throwing much out. Replacing the `.*` addresses the biggest issue. The point about associating a start tag with its end tag isn't a demand; it's an acknowledgement of how this stuff works unless we build far more complicated patterns (purely due to the recursive nature of HTML).